### PR TITLE
storage: add `mz_sink_statistics`

### PR DESCRIPTION
**Pull request by guswynn**
This is the `sink` parallel to https://github.com/MaterializeInc/materialize/pull/16788, that additionally adds linked-prometheus metrics, which are used to partially close: https://github.com/MaterializeInc/materialize/issues/15904

The added metrics are

- `messages_staged`/`bytes_staged`: the message/byte count of non-committed messages sinked into kafka
- `messages_committed`/`bytes_committed`: the message/byte count of committed messages

There are no bools like `snapshot_committed` for sources, as I believe there is no parallel in sinks? @bkirwi @aljoscha Ill need your help determining if this is a minimal viable set of metrics for sinks. 

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

This pr, minus the actual metrics bumps in `sink/kafka.rs`, is effectively all boiler-plate. I tried to share as much code as possible with the source stats stuff, but there was some amount of unavoidable copy-able.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Add `mz_sink_statistics`

### CODE REVIEW

File: src/adapter/src/catalog/builtin.rs [hunk 1]
```diff
@@ -1765,6 +1765,19 @@ pub static MZ_SOURCE_STATISTICS: Lazy<BuiltinSource> = Lazy::new(|| BuiltinSourc
...
     is_retained_metrics_relation: true,
 });
+pub static MZ_SINK_STATISTICS: Lazy<BuiltinSource> = Lazy::new(|| BuiltinSource {
+    name: "mz_sink_statistics",
+    schema: MZ_INTERNAL_SCHEMA,
+    data_source: Some(IntrospectionType::StorageSinkStatistics),
+    desc: RelationDesc::empty()
+        .with_column("id", ScalarType::String.nullable(false))
+        .with_column("worker_id", ScalarType::UInt64.nullable(false))
+        .with_column("messages_staged", ScalarType::UInt64.nullable(false))
```
Comment (aljoscha) [inline src/adapter/src/catalog/builtin.rs:1775]: the source metrics have these grouped differently: I think here it should also be:
 - `messages_staged`
 - `messages_committed`
 - `bytes_staged`
 - `bytes_committed`

File: src/storage-client/src/client.rs [hunk 2]
```diff
@@ -287,11 +287,13 @@ impl Arbitrary for StorageCommand<mz_repr::Timestamp> {
...
 }
 
-/// This structure represents a full set up updates for the `mz_source_statistics`
-/// table, for a specific source-worker pair. It is structured like this for simplicity
-/// and efficiency: Each storage worker can individually collect and consolidate metrics,
-/// then control how much `StorageResponse` traffic is produced when sending updates
-/// back to the controller to be written.
+// These structure represents a full set up updates for the `mz_source_statistics`
+// and `mz_sink_statistics` tables for a specific source-worker pair.
+// They are structured like this for simplicity
```
Comment (aljoscha) [inline src/storage-client/src/client.rs:292]: "source-worker/sink-worker" ?

File: src/storage-client/src/client.proto [hunk 3]
```diff
@@ -81,8 +81,17 @@ message ProtoStorageResponse {
         uint64 updates_committed = 6;
         uint64 bytes_received = 7;
     }
+    message ProtoSinkStatisticsUpdate {
+        mz_repr.global_id.ProtoGlobalId id = 1;
+        uint64 worker_id = 2;
+        uint64 messages_staged = 3;
```
Comment (aljoscha) [inline src/storage-client/src/client.proto:87]: this is very nit-y: but I'd also order them differently (see my comment above)

File: src/storage-client/src/client.rs [hunk 4]
```diff
@@ -303,6 +305,42 @@ pub struct SourceStatisticsUpdate {
     pub bytes_received: u64,
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct SinkStatisticsUpdate {
+    pub id: GlobalId,
+    pub worker_id: usize,
+    pub messages_staged: u64,
```
Comment (aljoscha) [inline src/storage-client/src/client.rs:312]: this is very nit-y: but I'd also order them differently (see my comment above)

File: src/storage-client/src/client.rs [hunk 5]
```diff
@@ -303,6 +305,42 @@ pub struct SourceStatisticsUpdate {
...
+    pub bytes_staged: u64,
+    pub messages_committed: u64,
+    pub bytes_committed: u64,
+}
+
+/// A trait that abstracts over user-facing statistics objects, used
+/// by `spawn_statistics_scraper`.
+pub trait PackableStats {
```
Comment (aljoscha) [inline src/storage-client/src/client.rs:320]: nice! ✊

File: src/storage-client/src/controller/statistics.rs [hunk 6]
```diff
@@ -75,11 +75,7 @@ pub(super) fn spawn_statistics_scraper(
 
                                 packer.push(Datum::from(src_id.to_string().as_str()));
                                 packer.push(Datum::from(u64::cast_from(*worker_id)));
-                                packer.push(Datum::from(stats.snapshot_committed));
-                                packer.push(Datum::from(stats.messages_received));
-                                packer.push(Datum::from(stats.updates_staged));
-                                packer.push(Datum::from(stats.updates_committed));
-                                packer.push(Datum::from(stats.bytes_received));
+                                stats.pack(packer);
```
Comment (aljoscha) [inline src/storage-client/src/controller/statistics.rs:78]: why not pack the worker and id in the `Packable` as well?

Comment (aljoscha) [inline src/storage-client/src/controller/statistics.rs:78]: I can't comment up above, but the loop variables should probably not be `src_id`/`sources` anymore?

File: src/storage/src/statistics.rs [hunk 7]
```diff
@@ -132,6 +134,77 @@ impl SourceStatisticsMetrics {
...
+    pub(crate) messages_staged: IntCounterVec,
+    pub(crate) bytes_staged: IntCounterVec,
+    pub(crate) messages_committed: IntCounterVec,
+    pub(crate) bytes_committed: IntCounterVec,
+}
+
+impl SinkStatisticsMetricsDefinitions {
+    pub(crate) fn register_with(registry: &MetricsRegistry) -> Self {
+        Self {
+            messages_staged: registry.register(metric!(
```
Comment (aljoscha) [inline src/storage/src/statistics.rs:148]: Taking a step back: these metrics names can be somewhat confusing now. There's `mz_updates_staged` (for sources), and `mz_messages_staged` (for sinks). Should we maybe prefix all of these with `mz_source_`/`mz_sink`?

File: src/storage/src/statistics.rs [hunk 8]
```diff
@@ -132,6 +134,77 @@ impl SourceStatisticsMetrics {
...
+                name: "mz_messages_staged",
+                help: "The number of messages staged but possibly not committed to the sink.",
+                var_labels: ["source_id", "worker_id"],
```
Comment (aljoscha) [inline src/storage/src/statistics.rs:151]: The labels should be `sink_id`.

File: src/storage/src/storage_state.rs [hunk 9]
```diff
@@ -293,16 +299,20 @@ impl<'w, A: Allocate> Worker<'w, A> {
 
                     // Initialize shared frontier tracking.
                     for (export_id, export) in ingestion.description.source_exports.iter() {
-                        self.storage_state.source_statistics.insert(
+                        // This is a separate line cause rustfmt :(
```
Comment (aljoscha) [inline src/storage/src/storage_state.rs:302]: 😞

File: test/sqllogictest/cluster_log_drop.slt [hunk 10]
```diff
@@ -23,7 +23,7 @@ reset-server
 query T rowsort
 SELECT (SELECT COUNT(*) FROM mz_catalog.mz_sources WHERE name NOT LIKE '%_1' AND name NOT LIKE '%_2' AND name NOT LIKE '%_3') - (SELECT COUNT(*) FROM mz_catalog.mz_sources WHERE name LIKE '%_1');
 ----
-5
+6
```
Comment (aljoscha) [inline test/sqllogictest/cluster_log_drop.slt:26]: These tests, one day... 🤜

**Review by aljoscha — COMMENTED** [review]
Nice work on (mostly) factoring out the common code! I did have some nits, and there where some copy-pasta things. Plus I had the larger question of what the naming of the source/sink metrics should be, to help avoid confusion.

File: src/storage/src/statistics.rs (hunk 7 above)
Comment (guswynn) [reply src/storage/src/statistics.rs:148]: yeah, I can clean these up!

File: src/storage/src/storage_state.rs (hunk 9 above)
Comment (guswynn) [reply src/storage/src/storage_state.rs:302]: I know...

File: test/sqllogictest/cluster_log_drop.slt (hunk 10 above)
Comment (guswynn) [reply test/sqllogictest/cluster_log_drop.slt:26]: I know!!

File: src/storage-client/src/client.rs (hunk 4 above)
Comment (guswynn) [reply src/storage-client/src/client.rs:312]: I agree, all the orderings should match across the entire codebase~

**Comment by guswynn** [comment]
Updated and fixed all the nits!

File: src/storage/src/statistics.rs [hunk 11]
```diff
@@ -235,3 +320,51 @@ impl SourceStatistics {
...
+impl StorageStatistics<SinkStatisticsUpdate, SinkStatisticsMetrics> {
+    pub(crate) fn new(id: GlobalId, worker_id: usize, metrics: &SinkBaseMetrics) -> Self {
+        Self {
+            stats: Rc::new(RefCell::new((
+                // We have no snapshot metrics for sinks as of now.
+                true,
+                SinkStatisticsUpdate {
+                    id,
+                    worker_id,
+                    messages_staged: 0,
```
Comment (aljoscha) [inline src/storage/src/statistics.rs:333]: very much a nit: ordering is still different :sweat_smile:

**Review by aljoscha — APPROVED** [review]
excllente!

**Review by jkosh44 — APPROVED** [review]
Surfaces owed files LGTM

File: src/storage/src/statistics.rs (hunk 11 above)
Comment (guswynn) [reply src/storage/src/statistics.rs:333]: the _one_ place I missed!

**Pull request merged**